### PR TITLE
Cloudinit files per buildagent and jumpboxes

### PR DESCRIPTION
**Pull request by sebader**
- Removes the need to install npm on the agents
- Different cloud init file for jumpboxes

### CODE REVIEW

**Comment by msimecek** [comment]
e2e?

**Review by msimecek — APPROVED** [review]
LGTM

**Comment by heoelri** [comment]
Good idea - I still struggle a bit from time to time when to install them (e.g. via cloud init) and when to install them via the pipeline - we'd the discussion a couple of times now e.g. for terraform. It definitely makese sense when we need to control the version and align it across all steps (w/o sed magic). But on the other hand could including them into the machine build/deploy process considered as potential improvements/speedup for the pipelines execution.

File: .ado/pipelines/templates/jobs-code-tests.yaml [hunk 1]
```diff
@@ -23,6 +23,11 @@ jobs:
       command: 'test'
       workingDirectory: ${{ parameters.workingDirectory }} # path to SLN
 
+  - task: NodeTool@0
+    displayName: 'Use Node v16 for UI app'
+    inputs:
+      versionSpec: '16.x'
```
Comment (heoelri) [inline .ado/pipelines/templates/jobs-code-tests.yaml:29]: Would it make sense to consolidate that into `./ado/config/configuration.yaml`?

File: .ado/pipelines/templates/jobs-ui-app-build.yaml [hunk 2]
```diff
@@ -8,6 +8,11 @@ jobs:
   displayName: 'Build UI app'
   steps:
 
+  - task: NodeTool@0
+    displayName: 'Use Node v16 for UI app'
+    inputs:
+      versionSpec: '16.x'
```
Comment (heoelri) [inline .ado/pipelines/templates/jobs-ui-app-build.yaml:14]: Especially given the fact that the version number is defined twice?

**Comment by sebader** [comment]
> Good idea - I still struggle a bit from time to time when to install them (e.g. via cloud init) and when to install them via the pipeline - we'd the discussion a couple of times now e.g. for terraform. It definitely makese sense when we need to control the version and align it across all steps (w/o sed magic). But on the other hand could including them into the machine build/deploy process considered as potential improvements/speedup for the pipelines execution.

this ask actually came from costco: they did some security scan and found an outdated nodejs version on the jumpbox. Thats the one that apt install nodjs via cloudinit would install. Thats why I removed it.

Ideally we can stay consistent on most of this stuff between hosted and private agents. I.e. only install on the private agents what is also preinstalled on the hosted agents

File: .ado/pipelines/templates/jobs-ui-app-build.yaml (hunk 2 above)
Comment (sebader) [reply .ado/pipelines/templates/jobs-ui-app-build.yaml]: great idea

Comment (sebader) [reply .ado/pipelines/templates/jobs-ui-app-build.yaml]: done

**Comment by heoelri** [comment]
> > Good idea - I still struggle a bit from time to time when to install them (e.g. via cloud init) and when to install them via the pipeline - we'd the discussion a couple of times now e.g. for terraform. It definitely makese sense when we need to control the version and align it across all steps (w/o sed magic). But on the other hand could including them into the machine build/deploy process considered as potential improvements/speedup for the pipelines execution.
> 
> this ask actually came from costco: they did some security scan and found an outdated nodejs version on the jumpbox. Thats the one that apt install nodjs via cloudinit would install. Thats why I removed it.
> 
> Ideally we can stay consistent on most of this stuff between hosted and private agents. I.e. only install on the private agents what is also preinstalled on the hosted agents

That's actually a pretty good approach I'd say. We're not publishing the list of packages our build agents contain, are we? I think we're doing that for github actions.. https://github.com/actions/virtual-environments

**Comment by heoelri** [comment]
We're publishing it.. https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md

**Comment by sebader** [comment]
jep. also good to go from your side now? @heoelri

**Review by heoelri — APPROVED** [review]

**Pull request merged**